cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -186,23 +189,39 @@ private File getUnpackLocalPath(ArtifactId artifactId, long lastModifiedTimestam
         .toFile();
   }
 
-  public void preloadArtifacts(Set<String> artifactNames)
+  /**
+   * Preloads the latest version of selected artifacts.
+   *
+   * @param artifactNames list of artifact names to be preloaded in cache
+   * @throws ArtifactNotFoundException if we couldnot find any artifacts for given name
+   */
+  public void preloadArtifacts(Set<String> artifactNames, int versionCount)
       throws IOException, ArtifactNotFoundException {
     ArtifactManager artifactManager = artifactManagerFactory.create(
         NamespaceId.SYSTEM,
         RetryStrategies.fromConfiguration(cConf, Constants.Service.TASK_WORKER + "."));
 
-    for (ArtifactInfo info : artifactManager.listArtifacts()) {
-      if (artifactNames.contains(info.getName()) && info.getParents().isEmpty()) {
-        String className = info.getClasses().getApps().stream()
-            .findFirst()
-            .map(ApplicationClass::getClassName)
-            .orElse(null);
 
-        LOG.info("Preloading artifact {}:{}-{}", info.getScope(), info.getName(),
-            info.getVersion());
+    List<ArtifactInfo> allArtifacts = artifactManager.listArtifacts();
+    for (String artifactName : artifactNames) {
+      List<ArtifactInfo> artifactsToPreload = allArtifacts.stream()
+          .filter(artifactInfo -> artifactInfo.getName().equals(artifactName))
+          .sorted(Comparator.comparing((ArtifactInfo artifactInfo) ->
+              new ArtifactVersion(artifactInfo.getVersion())).reversed())
+          .limit(versionCount)
+          .collect(Collectors.toList());
+
+      if (artifactsToPreload.isEmpty()) {
+          throw new ArtifactNotFoundException(NamespaceId.SYSTEM, artifactName);
```

Comment:
Do we really need to throw an exception? E.g., let's say we set version count to 0. Why throw an exception?

Reply:
As we generally want to preload the core system artifacts like wrangler-service and cdap-data-pipeline, my original idea was that if no matching artifacts are found that is unexpected. 
I think as the artifact list is config driven rather than failing a warn statement should be more apt.
Will change